lib/trie/trie.go

```diff
@@ -41,54 +39,54 @@ func NewTrie(root Node) *Trie {
 	}
 }
 
-// Snapshot created a copy of the trie.
-func (t *Trie) Snapshot() *Trie {
-	children := make(map[common.Hash]*Trie)
-	for h, c := range t.childTries {
-		children[h] = &Trie{
-			generation:  c.generation + 1,
-			root:        c.root,
+// Snapshot creates a copy of the trie.
+// Note it does not deep copy the trie, but will
+// copy on write as modifications are done on this new trie.
+// It does a snapshot of all child tries as well, and resets
+// the set of deleted hashes.
+func (t *Trie) Snapshot() (newTrie *Trie) {
+	childTries := make(map[common.Hash]*Trie, len(t.childTries))
+	for rootHash, childTrie := range t.childTries {
+		childTries[rootHash] = &Trie{
+			generation:  childTrie.generation + 1,
+			root:        childTrie.root.Copy(false),
 			deletedKeys: make(map[common.Hash]struct{}),
 		}
 	}
 
-	newTrie := &Trie{
+	return &Trie{
 		generation:  t.generation + 1,
 		root:        t.root,
-		childTries:  children,
+		childTries:  childTries,
 		deletedKeys: make(map[common.Hash]struct{}),
 	}
-
-	return newTrie
 }
 
-func (t *Trie) maybeUpdateGeneration(n Node) Node {
-	if n == nil {
-		return nil
-	}
-
-	// Make a copy if the generation is updated.
-	if n.GetGeneration() < t.generation {
-		// Insert a new node in the current generation.
-		const copyChildren = false
-		newNode := n.Copy(copyChildren)
-
-		newNode.SetGeneration(t.generation)
-
-		// Hash of old nodes should already be computed since it belongs to older generation.
-		oldNodeHash := n.GetHash()
-		if len(oldNodeHash) > 0 {
-			hash := common.BytesToHash(oldNodeHash)
-			t.deletedKeys[hash] = struct{}{}
-		}
-		return newNode
+// updateGeneration is called when the currentNode is from
+// an older trie generation (snapshot) so we deep copy the
+// node and update the generation on the newer copy.
+func updateGeneration(currentNode Node, trieGeneration uint64,
```

Comment:
should we leave the check for if the currentNode generation matches `trieGeneration`?

Reply:
So....

I tried the following:

1. Add a generation equal check and return `currentNode` if the generations are equal.. but then this is back to `maybeUpdateGeneration` 😄 so that feels wrong
2. Add a generation equal check and panic if it's equal. Since this code should be called when the generation differs only, and we really don't want to deep copy a node for no reason.

I'm planning sometime in the future to call `updateGeneration` only when the node is modified. Right now we call it (and hence deep copy) even if no modification is done downstream. That would save some time since we would only deep copy when it would be actually needed. So I will keep `2.` to help development of this enhancement, since even unit tests might not pick up the deep copying really.


I added a check with panic if the node generation and trie generation are the same, since the caller shouldn't be calling this code path really, and this is susceptible to happen with future changes since it's a nifty detail.